### ISSUE TITLE
FEATURE: More flexible id for NodeAggregate

### ISSUE THREAD

**Issue by dfeyer** [neos/neos-development-collection#5111]
### Is there an existing issue for this topic?

- [X] I have searched the existing issues

### Description

This issue is a follow-up on https://discuss.neos.io/t/rfc-non-uuid-node-identifiers/1997/16

It should be possible to use a more flexible ID for nodes. In many cases, Neos collaborate with external systems, which requires data synchronization. Being able to use arbitrary IDs can help a lot with this process.

### Possible Solution

An acceptable solution can be to loosen a bit the current regexp used in NodeAggregateId with the following one:

/^[a-zA-Z0-9-_.]{1,64}$/

- the limit of 64 can be discussed regarding index performance
- capital letters allow the support of base58 or nanoid
- dot and underscore are pretty common, by ex. ContentFull

**Comment by dfeyer** [comment]
a direct use case of this can be the ContentRepositoryImporter if I don't have to maintain the mapping between external ids and the node ids it can make the package more simple and I can replace the mapping by a simple id prefixer strategy to avoid collisions

**Comment by bwaidelich** [comment]
I agree that we can relax the constraints even further, but this can have unwanted effects to performance (DB schema) and rendering (e.g. if we allow "<" it can suddenly pose a new XSS thread). So we should carefully think about this.

When working with external services that have really really weird IDs, you can always use a hash (and store the original id in a node property if you need the reverse direction).

> /^[a-zA-Z0-9-_.]{1,64}$/

I think that's a range that we can live with – however, allowing lower _and_ upper case characters increases the risk of errors slightly

**Note:** We must not forget to adjust the length (and potentially type) of the [database schema](https://github.com/neos/neos-development-collection/blob/cd863719d49b1eb0a722769894d33e7de4102bb0/Neos.ContentRepository.Core/Classes/Infrastructure/DbalSchemaFactory.php#L38-L41) if we change that

**Comment by kitsunet** [comment]
What @bwaidelich said. I am hesitant to increase size due to DB indices and schema BUT I could live with 64 I guess.

**Comment by mhsdesign** [comment]
For reference the current pattern is `^([a-z0-9\-]{1,64})$` so in terms of length nothing changes.
But i see that youll proposed support for `_` `.` and uppercase `A-Z`

**Comment by bwaidelich** [comment]
BTW: Allowing the underscore to be part of the aggregate id would break our current implementation of [CacheTags](https://github.com/neos/neos-development-collection/blob/cffdca7861e235c6dd80c91011ebc311fd922449/Neos.Neos/Classes/Fusion/Cache/CacheTag.php#L62) (because we use an underscore to separate the cache parts).

I suggest to change that (see https://github.com/neos/neos-development-collection/pull/5122#discussion_r1625442310) but as of now that is the case

**Comment by mhsdesign** [comment]
Fyi we would also need to adjust `\Neos\Neos\Fusion\ConvertUrisImplementation::PATTERN_SUPPORTED_URIS` and `\Neos\Neos\Service\LinkingService::PATTERN_SUPPORTED_URIS`
